matchmake-extension/auto_matchmake_with_param_postpone.go

```diff
@@ -4,53 +4,82 @@ import (
 )
 
 func (commonProtocol *CommonProtocol) autoMatchmakeWithParamPostpone(err error, packet nex.PacketInterface, callID uint32, autoMatchmakeParam *match_making_types.AutoMatchmakeParam) (*nex.RMCMessage, *nex.Error) {
+	if commonProtocol.CleanupMatchmakeSessionSearchCriterias == nil {
+		common_globals.Logger.Warning("MatchmakeExtension::AutoMatchmakeWithParam_Postpone missing CleanupMatchmakeSessionSearchCriterias!")
+		return nil, nex.NewError(nex.ResultCodes.Core.NotImplemented, "change_error")
+	}
+
 	if err != nil {
 		common_globals.Logger.Error(err.Error())
 		return nil, nex.NewError(nex.ResultCodes.Core.InvalidArgument, "change_error")
 	}
 
+	if !common_globals.CheckValidMatchmakeSession(autoMatchmakeParam.SourceMatchmakeSession) {
+		return nil, nex.NewError(nex.ResultCodes.Core.InvalidArgument, "change_error")
+	}
+
+	if len(autoMatchmakeParam.JoinMessage.Value) > 256 {
+		return nil, nex.NewError(nex.ResultCodes.Core.InvalidArgument, "change_error")
+	}
+
 	connection := packet.Sender().(*nex.PRUDPConnection)
 	endpoint := connection.Endpoint().(*nex.PRUDPEndPoint)
 
+	commonProtocol.manager.Mutex.Lock()
+
 	// * A client may disconnect from a session without leaving reliably,
 	// * so let's make sure the client is removed from the session
-	common_globals.RemoveConnectionFromAllSessions(connection)
+	database.EndMatchmakeSessionsParticipation(commonProtocol.manager, connection)
 
-	matchmakeSession := autoMatchmakeParam.SourceMatchmakeSession
+	commonProtocol.CleanupMatchmakeSessionSearchCriterias(autoMatchmakeParam.LstSearchCriteria)
 
-	sessions := common_globals.FindSessionsByMatchmakeSessionSearchCriterias(connection.PID(), autoMatchmakeParam.LstSearchCriteria.Slice(), commonProtocol.GameSpecificMatchmakeSessionSearchCriteriaChecks)
-	var session *common_globals.CommonMatchmakeSession
+	resultRange := types.NewResultRange()
+	resultRange.Length.Value = 1
+	resultSessions, nexError := database.FindMatchmakeSessionBySearchCriteria(commonProtocol.manager, connection, autoMatchmakeParam.LstSearchCriteria.Slice(), resultRange)
+	if nexError != nil {
+		commonProtocol.manager.Mutex.Unlock()
+		return nil, nexError
+	}
 
-	if len(sessions) == 0 {
-		var errCode *nex.Error
-		session, errCode = common_globals.CreateSessionByMatchmakeSession(matchmakeSession, nil, connection.PID())
-		if errCode != nil {
-			common_globals.Logger.Error(errCode.Error())
-			return nil, errCode
+	var resultSession *match_making_types.MatchmakeSession
+	if len(resultSessions) == 0 {
+		resultSession = autoMatchmakeParam.SourceMatchmakeSession.Copy().(*match_making_types.MatchmakeSession)
+		nexError = database.CreateMatchmakeSession(commonProtocol.manager, connection, resultSession)
+		if nexError != nil {
+			common_globals.Logger.Error(nexError.Error())
+			commonProtocol.manager.Mutex.Unlock()
+			return nil, nexError
 		}
 	} else {
-		session = sessions[0]
+		resultSession = resultSessions[0]
+
+		// TODO - What should really happen here?
```

Comment:
If I may suggest:

There are actually two errors that exist in the error-codes repo: `RendezVous::MatchmakeSessionSystemPasswordUnmatch` and `RendezVous::MatchmakeSessionUserPasswordUnmatch`.

Maybe throw the error depending on which one is not what it should be?

Reply:
I don't think those are the right errors here since this happens in an `AutoMatchmake` method, so the user cannot specify a password anyway